### PR TITLE
Cleaning up titles for the hibernate rules.

### PR DESCRIPTION
**Pull request by Maarc**
Consolidated all hibernate rules. Added the deleted catchall back. Correct several titles.

Category corrected from "mandatory" to "optional" as it is not a mandatory change in a migration to EAP 6/7 : no hibernate/JPA update is mandatory to get a running application.

### CODE REVIEW

**Comment by windupgithubbot1** [comment]
Can one of the admins verify this patch?

File: rules-reviewed/hibernate/hibernate-catchall.windup.xml [hunk 1]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0"?>
+<ruleset xmlns="http://windup.jboss.org/schema/jboss-ruleset" id="hibernate-catchall" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (d-s) [inline rules-reviewed/hibernate/hibernate-catchall.windup.xml]: Looks like this was discussed and then removed in https://issues.jboss.org/browse/WINDUPRULE-135 -- will let Marek take a look, but we should discuss if this needs to be added back.

Comment (Maarc) [inline rules-reviewed/hibernate/hibernate-catchall.windup.xml]: The rules were too broad in WR-135 because the class names were used in the title of the rule ... generating a large amount of noise.

With those rules, the effort level is "0" (pure information) and all entries are bundled together.

Comment (d-s) [inline rules-reviewed/hibernate/hibernate-catchall.windup.xml]: I see what you're saying. I don't have a problem with adding it back this way.

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml [hunk 2]
```diff
@@ -515,7 +498,7 @@
                 <javaclass references="org.hibernate.metamodel.spi.{type}"/>
             </when>
             <perform>
-                <hint title="org.hibernate.metamodel.spi.{type} was moved in Hibernate 5" effort="1" category-id="mandatory">
+                <hint title="Hibernate 5 - Moved method org.hibernate.metamodel.spi.{type}" effort="1" category-id="optional">
```
Comment (mareknovotny) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: while deprecated methods could be **Optional**, package movements or class removal _are not_ **Optional**.

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml [hunk 3]
```diff
@@ -534,7 +517,7 @@
                 <javaclass references="org.hibernate.cfg.AnnotationConfiguration"/>
             </when>
             <perform>
-                <hint title="org.hibernate.cfg.AnnotationConfiguration was removed in Hibernate 5" effort="1" category-id="mandatory">
+                <hint title="Hibernate 5 - Removed class org.hibernate.cfg.AnnotationConfiguration" effort="1" category-id="optional">
```
Comment (mareknovotny) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: while deprecated methods could be *Optional*, package movements or class removal are not *Optional*.

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml [hunk 4]
```diff
@@ -553,7 +536,7 @@
                 <javaclass references="org.hibernate.id.{idclass}"/>
             </when>
             <perform>
-                <hint title="Class org.hibernate.id.{idclass} was removed/deprecated in Hibernate 5" effort="1" category-id="mandatory">
+                <hint title="Hibernate 5 - Removed deprecated class org.hibernate.id.{idclass}" effort="1" category-id="optional">
```
Comment (mareknovotny) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: while deprecated methods could be *Optional*, package movements or class removal are not *Optional*.

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml [hunk 5]
```diff
@@ -573,7 +556,7 @@
                 <javaclass references="org.hibernate.hql.spi.{type}" />
             </when>
             <perform>
-                <hint title="Class org.hibernate.hql.spi.{type} was moved in Hibernate 5" effort="1" category-id="mandatory">
+                <hint title="Hibernate 5 - Moved class org.hibernate.hql.spi.{type}" effort="1" category-id="optional">
```
Comment (mareknovotny) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: again package movements or class removal are not *Optional*.

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml [hunk 6]
```diff
@@ -590,7 +573,7 @@
                 <javaclass references="org.hibernate.hql.spi.PersistentTableBulkIdStrategy" />
             </when>
             <perform>
-                <hint title="Class org.hibernate.hql.spi.PersistentTableBulkIdStrategy was moved in Hibernate 5" effort="1" category-id="mandatory">
+                <hint title="Hibernate 5 - Moved class org.hibernate.hql.spi.PersistentTableBulkIdStrategy" effort="1" category-id="optional">
```
Comment (mareknovotny) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: again package movements or class removal are not *Optional*.

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml [hunk 7]
```diff
@@ -604,7 +587,7 @@
                 <javaclass references="org.hibernate.hql.spi.TemporaryTableBulkIdStrategy" />
             </when>
             <perform>
-                <hint title="Class org.hibernate.hql.spi.TemporaryTableBulkIdStrategy was replaced in Hibernate 5" effort="1" category-id="mandatory">
+                <hint title="Hibernate 5 - Replaced class org.hibernate.hql.spi.TemporaryTableBulkIdStrategy" effort="1" category-id="optional">
```
Comment (mareknovotny) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: again package movements or class removal are not *Optional*.

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml [hunk 8]
```diff
@@ -620,7 +603,7 @@
                 </javaclass>
             </when>
             <perform>
-                <hint title="Interface org.hibernate.integrator.spi.Integrator was changed in Hibernate 5" effort="1" category-id="optional">
+                <hint title="Hibernate 5 - Changed interface org.hibernate.integrator.spi.Integrator" effort="1" category-id="optional">
```
Comment (mareknovotny) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: This is IMHO mandatory change too -> Changed interface

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml [hunk 9]
```diff
@@ -660,7 +643,7 @@
                 </javaclass>
             </when>
             <perform>
-                <hint title="Interface org.hibernate.id.Configurable changed method parameter in configure method in Hibernate 5" effort="1" category-id="mandatory">
+                <hint title="Hibernate 5 - Changed signature org.hibernate.id.Configurable" effort="1" category-id="optional">
```
Comment (mareknovotny) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: This is IMHO mandatory change too -> Changed interface

File: rules-reviewed/eap7/eap6/hsearch.windup.xml [hunk 10]
```diff
@@ -21,7 +21,7 @@
                 <javaclass references="org.hibernate.search.{*}" />
             </when>
             <perform>
-                <classification title="Hibernate Search API usage" effort="0" category-id="mandatory">
+                <classification title="Hibernate Search - API usage" effort="0" category-id="optional">
```
Comment (mareknovotny) [inline rules-reviewed/eap7/eap6/hsearch.windup.xml]: ok, classifications are perhaps by its usage all Optional. No problem with this.

File: rules-reviewed/eap7/eap6/hsearch.windup.xml [hunk 11]
```diff
@@ -32,7 +32,7 @@
                 <javaclass references="org.hibernate.search.impl.SearchMappingBuilder" />
             </when>
             <perform>
-                <hint title="Hibernate Search 5 renamed class SearchMappingBuilder" effort="1" category-id="mandatory">
+                <hint title="Hibernate Search 5 - Renamed class SearchMappingBuilder" effort="1" category-id="optional">
```
Comment (mareknovotny) [inline rules-reviewed/eap7/eap6/hsearch.windup.xml]: This is IMHO mandatory change too -> Renamed class.

File: rules-reviewed/hibernate/hibernate-catchall.windup.xml (hunk 1 above)
Comment (mareknovotny) [inline rules-reviewed/hibernate/hibernate-catchall.windup.xml]: I don't mind to add them, but only deprecated things can be **Optional** and removed/moved/declaration changes needs to be **Mandatory** IMO. If you will migrate you have to change the affected code.

File: rules-reviewed/hibernate/hibernate.windup.xml [hunk 12]
```diff
@@ -22,7 +22,7 @@
             <javaclass references="org.hibernate.annotations.Type" matchesSource="{*}text{*}" />
          </when>
          <perform>
-            <hint title="@Type(type=text) is deprecated in Hibernate 4" effort="1" category-id="mandatory">
+            <hint title="Hibernate 4 - Deprecated @Type(type=text)" effort="1" category-id="optional">
```
Comment (mareknovotny) [inline rules-reviewed/hibernate/hibernate.windup.xml]: The rewording of titles to take the last word and move it to the start is not really significant - this is not related to only this title change as here is a category-id change too, but wherever is only this rewording that is the matter of taste - not really additional value ;)

File: rules-reviewed/hibernate/hibernate.windup.xml [hunk 13]
```diff
@@ -176,8 +174,8 @@
             <javaclass references="net.sf.hibernate.{classname}" />
          </when>
          <perform>
-            <classification title="Hibernate 2.x" category-id="mandatory" />
-            <hint effort="1" category-id="mandatory" title="Hibernate type reference net.sf.hibernate.{classname}">
+            <classification title="Hibernate 2.x - Renamed package 'net.sf.hibernate'" category-id="optional" />
+            <hint effort="1" category-id="optional" title="Hibernate type reference net.sf.hibernate.{classname}">
```
Comment (mareknovotny) [inline rules-reviewed/hibernate/hibernate.windup.xml]: Not an Optional thing

File: rules-reviewed/hibernate/hibernate.windup.xml [hunk 14]
```diff
@@ -165,7 +163,7 @@
             <javaclass references="org.hibernate.collection.PersistentBag" />
          </when>
          <perform>
-            <hint title="Replace class org.hibernate.collection.PersistentBag" effort="1" category-id="mandatory">
+            <hint title="Hibernate 4 - Replace org.hibernate.collection.PersistentBag" effort="1" category-id="optional">
```
Comment (mareknovotny) [inline rules-reviewed/hibernate/hibernate.windup.xml]: Not an optional thing

File: rules-reviewed/hibernate/hibernate-catchall.windup.xml (hunk 1 above)
Comment (Maarc) [inline rules-reviewed/hibernate/hibernate-catchall.windup.xml]: I think that we have a small misunderstanding.

While conducting a hibernate update from 3 to 4 or from 4 to 5, you are perfectly right regarding "optional" and "mandatory" rules.

In an EAP migration project, updating Hibernate to a newer version is optional, as hibernate can be embedded in the application and disabled at application server-level. Considering a minimalistic migration (doing as few changes as possible), this is the preferred approach: we keep an embedded version of Hibernate in the application and do not touch the source code.

Comment (mareknovotny) [inline rules-reviewed/hibernate/hibernate-catchall.windup.xml]: Well, that is the different view on it and I accept that as possible way. 

The thing is that migration rules in core rulesets should support/provide the recommended migration path not minimalistic path. Because It depends on migrated application and team focus I think.
There could be migration focused on just for hibernate and then it will fail with minimalistic approach, but I get what you are trying to say.

Comment (jsight) [inline rules-reviewed/hibernate/hibernate-catchall.windup.xml]: I'm -1 on switching these to just "Mandatory". However, might it be possible to use a new issue category for these?

Perhaps "Mandatory for Hibernate Upgrade" or something along those lines (preferably something less verbose)? It is certainly a valid point that someone could migrate an application without changing these.

Comment (mareknovotny) [inline rules-reviewed/hibernate/hibernate-catchall.windup.xml]: @jsight it is a good suggestion IMO.

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml (hunk 2 above)
Comment (OndraZizka) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: According to Marc, this is optional because the common practice is to disable Hibernate 5 and add legacy Hibernate. This should also be explained in the hint.

Comment (OndraZizka) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: Is there some official documentation specifically for this replacing of Hibernate?

File: rules-reviewed/hibernate/hibernate-catchall.windup.xml (hunk 1 above)
Comment (Maarc) [inline rules-reviewed/hibernate/hibernate-catchall.windup.xml]: Please leave those as optional as suggested. In many cases, upgrading to the latest available JPA / Hibernate version is a very bad idea as it implies high efforts (code rewrite) and a lot of testing. For applications being at the end of their lifecycle, this is definitely not an option.

The "true" recommendation for applications using hibernate is however missing: keeping the used embedded ORM as-is and disabling JPA in EAP 6 / 7.

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml (hunk 2 above)
Comment (Maarc) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: @mareknovotny switching to another version of hibernate/JPA is definitely optional while moving to EAP 6/7.

There is no real direct knowledge article explaining that. Here is what I tend to recommend when I do a review:

RATIONALE
------------
One or several object-relational mapping (ORM) frameworks have been detected within the application (eclipselink, eclipselink-jpa, hibernate, hibernate-jpa, lift-jpa, scala-jpa, spring-data-jpa, ...) These frameworks may deviate from the out-of-the-box Java EE supported ORM in JBoss EAP.

RECOMMENDATION
---------------------
All used ORM frameworks should be listed and analyzed more thoroughly to prevent conflicts.

No update to the latest Hibernate or JPA version is required. The currently used ORM frameworks can remain embedded within the application while disabling the EAP out-of-the-box ORM using explicit class loading dependencies (jboss-deployment-structure.xml).
At the same time, an enterprise architecture decision might be taken to update or replace all ORM frameworks by the latest JPA standard. Only the version of Hibernate shipped with the EAP release is fully supported.

REFERENCES
--------------
https://access.redhat.com/solutions/306493
https://access.redhat.com/solutions/707243

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml (hunk 3 above)
Comment (Maarc) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: See comment above.

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml (hunk 4 above)
Comment (Maarc) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: See comment above.

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml (hunk 5 above)
Comment (Maarc) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: See comment above.

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml (hunk 6 above)
Comment (Maarc) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: See comment above.

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml (hunk 7 above)
Comment (Maarc) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: See comment above.

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml (hunk 8 above)
Comment (Maarc) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: See comment above.

File: rules-reviewed/eap7/eap6/hibernate4.windup.xml (hunk 9 above)
Comment (Maarc) [inline rules-reviewed/eap7/eap6/hibernate4.windup.xml]: See comment above.

File: rules-reviewed/eap7/eap6/hsearch.windup.xml (hunk 11 above)
Comment (Maarc) [inline rules-reviewed/eap7/eap6/hsearch.windup.xml]: See comment above.

File: rules-reviewed/hibernate/hibernate.windup.xml (hunk 12 above)
Comment (Maarc) [inline rules-reviewed/hibernate/hibernate.windup.xml]: I disagree. This is not only about shifting the last word at the beginning. It is about consistency, conciseness, and simplicity of the titles. Without this reformating, the reports look like a mess.

File: rules-reviewed/hibernate/hibernate.windup.xml (hunk 14 above)
Comment (Maarc) [inline rules-reviewed/hibernate/hibernate.windup.xml]: See comment above.

File: rules-reviewed/hibernate/hibernate.windup.xml (hunk 13 above)
Comment (Maarc) [inline rules-reviewed/hibernate/hibernate.windup.xml]: See comment above.

**Comment by Maarc** [comment]
We discussed what I just wrote in several calls several weeks ago and the changes are not merged. I took the time to update and explain again the requested changes.

Is there something preventing to merge those changes to the master branch?

**Comment by mareknovotny** [comment]
@Maarc if we change the category to Optional we must include changes in hint text to explain how to disable JPA and include old hibernate libs into application libraries. This is not here in the PR, so it can't be merged as it is now.

**Comment by OndraZizka** [comment]
Rebased, reviewed, about to merge when the tests are ok

**Comment by OndraZizka** [comment]
Merged. Thanks Marc for the effort and patience! :)
